File: app/schemas/horoscope.py
from datetime import date
from typing import Literal

from pydantic import Field

from app.schemas.chart import CamelModel, NatalChartRequest


class YearlyHoroscopeRequest(NatalChartRequest):
    year: int = Field(ge=1900, le=2100)


class MonthlyHoroscopeRequest(YearlyHoroscopeRequest):
    month: int = Field(ge=1, le=12)


class HoroscopeWindow(CamelModel):
    start_date: date = Field(alias="startDate")
    end_date: date = Field(alias="endDate")
    label: str


class YearlyMonthOverview(CamelModel):
    month: int
    title: str
    focus_areas: list[str] = Field(alias="focusAreas")
    intensity_score: int = Field(alias="intensityScore")
    top_theme: str = Field(alias="topTheme")
    lucky_window: HoroscopeWindow = Field(alias="luckyWindow")
    caution_window: HoroscopeWindow = Field(alias="cautionWindow")


class YearlyHoroscopeResponse(CamelModel):
    year: int
    profile_summary: str = Field(alias="profileSummary")
    months: list[YearlyMonthOverview]


class HoroscopeSections(CamelModel):
    career: str
    money: str
    love: str
    risk: str


class HoroscopeDateInsight(CamelModel):
    date: date
    label: str
    reason: str


class HoroscopeEvidence(CamelModel):
    date: date
    headline: str
    detail: str
    tone: Literal["supportive", "challenging"]


class MonthlyHoroscopeResponse(CamelModel):
    year: int
    month: int
    summary: str
    sections: HoroscopeSections
    lucky_dates: list[HoroscopeDateInsight] = Field(alias="luckyDates")
    caution_dates: list[HoroscopeDateInsight] = Field(alias="cautionDates")
    evidence: list[HoroscopeEvidence]
    llm_enhanced: bool = Field(alias="llmEnhanced")


class MonthlyHoroscopeLLMDateInsight(CamelModel):
    label: str
    reason: str


class MonthlyHoroscopeLLMEvidence(CamelModel):
    headline: str
    detail: str


class MonthlyHoroscopeLLMResponse(CamelModel):
    summary: str
    sections: HoroscopeSections
    lucky_dates: list[MonthlyHoroscopeLLMDateInsight] = Field(alias="luckyDates")
    caution_dates: list[MonthlyHoroscopeLLMDateInsight] = Field(alias="cautionDates")
    evidence: list[MonthlyHoroscopeLLMEvidence]

File: app/lib/horoscope.ts
import type { Route } from "next";

export interface HoroscopeSearchParams {
  birthDate?: string;
  birthTime?: string;
  city?: string;
  country?: string;
  timezone?: string;
  year?: string;
  month?: string;
  timeUnknown?: string;
}

export interface HoroscopeWindow {
  startDate: string;
  endDate: string;
  label: string;
}

export interface YearlyHoroscopeMonth {
  month: number;
  title: string;
  focusAreas: string[];
  intensityScore: number;
  topTheme: string;
  luckyWindow: HoroscopeWindow;
  cautionWindow: HoroscopeWindow;
}

export interface YearlyHoroscopeResponse {
  year: number;
  profileSummary: string;
  months: YearlyHoroscopeMonth[];
}

export interface HoroscopeSectionSet {
  career: string;
  money: string;
  love: string;
  risk: string;
}

export interface HoroscopeDateInsight {
  date: string;
  label: string;
  reason: string;
}

export interface HoroscopeEvidence {
  date: string;
  headline: string;
  detail: string;
  tone: "supportive" | "challenging";
}

export interface MonthlyHoroscopeResponse {
  year: number;
  month: number;
  summary: string;
  sections: HoroscopeSectionSet;
  luckyDates: HoroscopeDateInsight[];
  cautionDates: HoroscopeDateInsight[];
  evidence: HoroscopeEvidence[];
  llmEnhanced: boolean;
}

export function hasRequiredBirthDetails(searchParams: HoroscopeSearchParams): boolean {
  return Boolean(
    searchParams.birthDate &&
      searchParams.birthTime &&
      searchParams.city &&
      searchParams.country
  );
}

export function getSelectedHoroscopeYear(
  searchParams: HoroscopeSearchParams,
  referenceDate: Date = new Date()
): number {
  const rawYear = Number(searchParams.year);
  if (Number.isInteger(rawYear) && rawYear >= 1900 && rawYear <= 2100) {
    return rawYear;
  }

  return referenceDate.getFullYear();
}

export function getSelectedHoroscopeMonth(
  selectedYear: number,
  requestedMonth?: string,
  referenceDate: Date = new Date()
): number {
  const parsedMonth = Number(requestedMonth);
  if (Number.isInteger(parsedMonth) && parsedMonth >= 1 && parsedMonth <= 12) {
    return parsedMonth;
  }

  if (selectedYear === referenceDate.getFullYear()) {
    return referenceDate.getMonth() + 1;
  }

  return 1;
}

export function buildHoroscopeApiParams(
  searchParams: HoroscopeSearchParams,
  selectedYear: number,
  selectedMonth?: number
): URLSearchParams {
  const params = new URLSearchParams({
    birthDate: searchParams.birthDate ?? "",
    birthTime: searchParams.birthTime ?? "",
    city: searchParams.city ?? "",
    countryCode: searchParams.country ?? "",
    year: String(selectedYear)
  });

  if (selectedMonth !== undefined) {
    params.set("month", String(selectedMonth));
  }

  if (searchParams.timezone) {
    params.set("timezone", searchParams.timezone);
  }

  return params;
}

export function buildHoroscopePageHref(
  searchParams: HoroscopeSearchParams,
  selectedYear: number,
  selectedMonth: number
): Route {
  const params = new URLSearchParams({
    birthDate: searchParams.birthDate ?? "",
    birthTime: searchParams.birthTime ?? "",
    city: searchParams.city ?? "",
    country: searchParams.country ?? "",
    year: String(selectedYear),
    month: String(selectedMonth),
    timeUnknown: searchParams.timeUnknown ?? "false"
  });

  if (searchParams.timezone) {
    params.set("timezone", searchParams.timezone);
  }

  return `/horoscope?${params.toString()}` as Route;
}

export function formatHoroscopeDate(value: string): string {
  const date = new Date(`${value}T00:00:00`);
  return `${date.getMonth() + 1}월 ${date.getDate()}일`;
}
